### ISSUE TITLE
Re-emergence of docword bug

### ISSUE THREAD

**Issue by davclark** [BIDData/BIDMach#27]
It seems that the `scripts/getdata.sh` script has resurfaced this old issue with a bad path. I think this is an easy fix?

**Comment by jcanny** [comment]
Actually from looking at your command output, it looks like a different problem. Could you please attach the output? It seems to be hanging at the end of the first tparse (getuci.sh line 36). In other words, its having trouble writing using the zlib compression library. It was able to read uncompressed files though, so its either a library linkage/version problem with zlib or a filesystem protection problem. 

I'm afraid its working fine on my similar-vintage Mac.

**Comment by davclark** [comment]
Sorry - I was juggling too many mental balls earlier when you were asking me about Xcode. But I was logged in remotely to my linux (Ubuntu 14.10) workstation (using a 1.0.0 tarball I just downloaded today). Is it possible there's a bug there? Here's a partial copy of where things got hung up:

```
Loading nips data
--2015-02-20 14:41:47--  http://archive.ics.uci.edu/ml/machine-learning-databases/bag-of-words/vocab.nips.txt
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.1.95
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.1.95|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 100807 (98K) [text/plain]
Saving to: ‘vocab.nips.txt’

100%[======================================>] 100,807     --.-K/s   in 0.06s   

2015-02-20 14:41:48 (1.60 MB/s) - ‘vocab.nips.txt’ saved [100807/100807]

Uncompressing docword.nips.txt.gz
746316 lines processed
^C
```

I also installed libjzlib-java to see if that would help, but the script still stops at the same point (zlib1g and zlib1g-dev were installed already).

**Comment by jcanny** [comment]
This might be a nasty one to debug. The output:

|746316 lines processed|

means that the doc tokenizer completed successfully but apparently wasnt 
able to write out its output (that's usually why it hangs). The output 
is compressed with zlib and that's the first time it uses zlib which 
makes me wonder if that's the problem.  I havent had that specific 
problem before, but that *kind* of problem is usually a lib problem *or* 
an access control problem. e.g. if you were logged in as a different 
user from the user that owns that directory.

A few other guesses: it may have actually written the output for the 
first parse, but was unable to load the vocab file for the second parse 
(maybe file protection or lock being held by an interrupted http call?). 
You can check by seeing if BIDMach/data/uci/nips/ contains anything 
other than vocab.txt

I tried on my own mac, and unfortunately it worked fine. Its probably 
easier to debug face-to-face, perhaps after the next meeting?

-John

**Comment by davclark** [comment]
I do see a number of gzip'd files there:

```
6.6K cols.imat.gz
8.2M docword.txt
2.2M docword.txt.gz
1.2M rows.imat.gz
374K vals.fmat.gz
 99K vocab.txt
```

I can also give you (or any other BIDMach developer) an account on this machine if desired.

The other thing is that I installed Java 8 since it sounded like that was a better future-proof foundation. Could that be a problem?

**Comment by jcanny** [comment]
I thought it was on your Mac? If not I need to know the platform and OS 
version (as there should be with any bug report).

Java 8 is completely untested. The prereq for BIDMach is Java 7 as in 
the installation guide. Java 8 might work, but your on your own.

But this is happening in the C/shell code, so Java isnt the problem.

It looks like the parser completed successfully on the first file, but 
couldnt start on the second one. The second file "vocab.txt" is there 
though. Could you give me a complete "ls -l" on that directory showing 
protections and ownership?

-John

**Comment by davclark** [comment]
Sorry - I should have highlighted this more clearly on my comment from Friday. The system is

 - Ubuntu 14.10 amd64  (on an i7 processor)
 - NVIDIA GTX 760 (GK104) running nouveau open-source drivers

In any case, I think I misunderstood you when we talked about the state of java 7 vs. 8 a few weeks ago - and I'll be sure to change that back (even though it's not related to the C/shell code). I'll poke around with this a bit more on Monday (tomorrow).

**Comment by jcanny** [comment]
Hi Dav,
It would be good if you could give me an account on that machine, so I 
can get things set up for Weds.

Some quick notes:
* I still need the full directory listing for the nips data directory to 
be able to figure out why it stalled - or I can try running the script 
itself.
* BIDMach requires java 7 right now, java 8 is untested.
* The latest BIDMach requires CUDA 6.5. Older versions used 5.5. They're 
not compatible.
* CUDA will not run on nouveau drivers, but when you installed CUDA it 
should have installed a new driver.

-John

**Comment by davclark** [comment]
I've reset the symlink on my machine back to jdk1.7.0, and will switch to the Nvidia drivers. I suspect that the recent upgrade to Ubuntu 14.10 reverted to nouveau drivers. I'm currently planning to just use the JCuda drivers distributed in the large BIDMach tarball.

Here's a full `ls` for now:

```
BIDMach_1.0.0-full-linux-x86_64⟫ ls -la data/uci/nips/
total 12180
drwxrwxr-x 2 dav dav    4096 Feb 20 14:44 .
drwxrwxr-x 3 dav dav    4096 Feb 20 12:10 ..
-rw-rw-r-- 1 dav dav    6711 Feb 20 14:44 cols.imat.gz
-rw-rw-r-- 1 dav dav 8509193 Feb 20 14:44 docword.txt
-rw-rw-r-- 1 dav dav 2288169 Feb 28  2008 docword.txt.gz
-rw-rw-r-- 1 dav dav 1164935 Feb 20 14:44 rows.imat.gz
-rw-rw-r-- 1 dav dav  382918 Feb 20 14:44 vals.fmat.gz
-rw-rw-r-- 1 dav dav  100807 Feb 28  2008 vocab.txt
```

I made an account for you on this machine. I'll send you credentials via slack, or just let me know if there's a different way that's more convenient to you.

**Comment by jcanny** [comment]
OK, the data pulling problem was due to a missing format file. It was 
always in the github repo, but I didnt put it in the (rather ad-hoc) 
packaging script that I use. Its there now, and I'll put it in all the 
zip files.

I dont think java is technically installed, at least ubuntu doesnt seem 
to know about it. the JVMs are in /opt but to put one in the default 
path, I think you need to do:

|sudo update-alternatives--config java

If you wont want to default to Java 7 for everyone, we can skip this step and tell people  to add /opt/jdk/bin to their path otherwise.

Yes, you always need to reinstall third-party drivers after a kernel update.

Note, the CUDA driver (which includes a display driver) is separate from the JCuda libraries. You need both.
A driver is different from a library in that it needs to be installed to make the kernel aware of it.

The scripts and binaries seem to be working and bidmach is working minus CUDA functionality.

-John
|

**Comment by davclark** [comment]
FYI - I've used a PPA to (re)install oracle Java 7, so it's the system default now. Also used the Ubuntu installer for CUDA 6.5 (it was for 14.04, but appears to have worked - I used the "RUN" version, not the "DEB"). Now getting ~11gf/sec on some sample code, which is comparable with what you post on the wiki. 

Just let me know when you've got the updated scripts for fetching data, and I'll re-install the tarball. I figure it's good to have the full set of sample data around. I'd consider the issue closed, but I'll leave it to you if you want to wait until you've updated the packaging script.

Thanks a lot for all the help!
